pennylane/workflow/_capture_qnode.py

```diff
@@ -97,26 +97,26 @@ def _get_qnode_prim():
     qnode_prim = jax.core.Primitive("qnode")
     qnode_prim.multiple_results = True
 
-    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-arguments, unused-argument
     @qnode_prim.def_impl
     def _(*args, qnode, shots, device, qnode_kwargs, qfunc_jaxpr, n_consts, batch_dims=None):
+        if shots != device.shots:
+            raise NotImplementedError(
+                "override shots are not yet supported with the program capture execution."
+            )
+        if qnode_kwargs["diff_method"] not in {"backprop", "best"}:
+            raise NotImplementedError(
+                "only backpropagation derivatives are supported at this time."
+            )
+
         consts = args[:n_consts]
         non_const_args = args[n_consts:]
 
-        def qfunc(*inner_args):
-            return jax.core.eval_jaxpr(qfunc_jaxpr, consts, *inner_args)
-
-        qnode = qml.QNode(qfunc, device, **qnode_kwargs)
-
-        if batch_dims is not None:
-
-            # pylint: disable=protected-access
-            return jax.vmap(partial(qnode._impl_call, shots=shots), batch_dims[n_consts:])(
-                *jax.tree_util.tree_leaves(non_const_args)
-            )
-
-        # pylint: disable=protected-access
-        return qnode._impl_call(*non_const_args, shots=shots)
+        if batch_dims is None:
+            return device.eval_jaxpr(qfunc_jaxpr, consts, *non_const_args)
+        return jax.vmap(partial(device.eval_jaxpr, qfunc_jaxpr, consts), batch_dims[n_consts:])(
+            *jax.tree_util.tree_leaves(non_const_args)
```

Comment:
Sorry, I might just lack some experience here. Why do we need the additional `tree_leaves` unpacking as compared to iterable unpacking in the non-vmap case? :thinking:

Reply:
@PietropaoloFrisoni

Reply:
This is in case the argument to a `qnode` is a `pytree` and `jax.vmap` is used. To be more precise, without the `jax.tree_util.tree_leaves` call, [this test](https://github.com/PennyLaneAI/pennylane/blob/ecc8fec95a6e324bb781e260a6e6692124400c8a/tests/capture/test_capture_qnode.py#L866) would not work.

Now that I notice, I can also replicate a similar issue with the following code without `jax.vmap`:

```
@qml.qnode(qml.device("default.qubit", wires=1))
def circuit(x, y):
    qml.RX(jax.numpy.pi * x["arr"][0] + y - y, wires=0)
    qml.RY(x["arr"][1] ** 2, wires=0)
    qml.RX(x["arr"][1] * x["arr"][2], wires=0)
    return qml.expval(qml.PauliZ(0))

y = jax.numpy.pi
x = {
    "arr": jax.numpy.array(
        [
            [0.1, 0.2, 0.3],
            [0.4, 0.5, 0.6],
            [0.7, 0.8, 0.9],
        ]
    ),
    "foo": None,
}

jaxpr = jax.make_jaxpr(circuit)(x, y)
result = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x, y)
```

That generates:

```
TypeError: Value {'arr': Traced<ShapedArray(float32[3,3])>with<DynamicJaxprTrace(level=1/0)>, 'foo': None} with type <class 'dict'> is not a valid JAX type
```

We could fix this replacing `*inner_args` with `*jax.tree_util.tree_leaves(inner_args)` [here](https://github.com/PennyLaneAI/pennylane/blob/ecc8fec95a6e324bb781e260a6e6692124400c8a/pennylane/workflow/_capture_qnode.py#L107).

Maybe there are other methods, so I am open to discussion 👍

Reply:
Edit: This is an even simpler case that can be fixed with the same trick:

```
import pennylane as qml
import jax

qml.capture.enable()

@qml.qnode(qml.device("default.qubit", wires=1))
def circuit(x):
    qml.RX(x["val"], 0)
    return qml.expval(qml.PauliZ(0))

x = {"val": 0.5}
jaxpr = jax.make_jaxpr(circuit)(x)
result = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x)
```

Reply:
Nice, this makes a lot of sense! In that case I'd request that we add the `tree_leaves` call to the non-vmap branch :)

Reply:
So why is this line insufficient:

https://github.com/PennyLaneAI/pennylane/blob/23faad982490cac46a1a1b025b976169d5a87f31/pennylane/workflow/_capture_qnode.py#L298

We should be flattening all the arguments before we hit `qnode_prim.bind` 🤔

Reply:
As discussed offline (thanks @albi3ro for figuring this out), the line Christina mentioned should be enough.

It is only necessary in the example above because we are calling:

```
result = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x)
```
Instead of 
```
result = jax.core.eval_jaxpr(jaxpr.jaxpr, jaxpr.consts, x["val"])
```

Reply:
Wasn't the result of this conversation that we can remove the `tree_leaves` from the code that was commented here? :thinking:

Reply:
Will get fixed in an independent PR.